collection_mutation.cc

```diff
@@ -282,16 +279,37 @@ collection_mutation merge(const abstract_type& type, collection_mutation_view a,
     };
 
     collection_mutation_view_description merged;
-    merged.cells.reserve(a_view.cells.size() + b_view.cells.size());
+    merged.cells.reserve(a.cells.size() + b.cells.size());
 
-    combine(a_view.cells.begin(), std::remove_if(a_view.cells.begin(), a_view.cells.end(), cell_killed(b_view.tomb)),
-            b_view.cells.begin(), std::remove_if(b_view.cells.begin(), b_view.cells.end(), cell_killed(a_view.tomb)),
+    combine(a.cells.begin(), std::remove_if(a.cells.begin(), a.cells.end(), cell_killed(b.tomb)),
+            b.cells.begin(), std::remove_if(b.cells.begin(), b.cells.end(), cell_killed(a.tomb)),
             std::back_inserter(merged.cells),
             compare,
             merge);
-    merged.tomb = std::max(a_view.tomb, b_view.tomb);
+    merged.tomb = std::max(a.tomb, b.tomb);
 
-    return merged.serialize(type);
+    return merged;
+}
+
+collection_mutation merge(const abstract_type& type, collection_mutation_view a, collection_mutation_view b) {
+    return a.with_deserialized(type, [&] (collection_mutation_view_description a_view) {
+    return b.with_deserialized(type, [&] (collection_mutation_view_description b_view) {
+
+    return visit(type, make_visitor(
+        [&] (const collection_type_impl& ctype) {
+            return merge(std::move(a_view), std::move(b_view), [&key_type = *ctype.name_comparator()] (bytes_view k1, bytes_view k2) {
+                return key_type.less(k1, k2);
+            });
+        },
+        [&] (const user_type_impl& utype) {
+            return merge(std::move(a_view), std::move(b_view), [] (bytes_view k1, bytes_view k2) {
+                return deserialize_field_index(k1) < deserialize_field_index(k2);
```

Comment:
We could avoid this by using short_type, no? That would limit UDTs to 32768 fields, but that should be acceptable.

Reply:
True, using uint16_t was my mistake anyway.